### PR TITLE
Feature/code refactoring

### PR DESCRIPTION
**Pull request by YauheniPo**

### CODE REVIEW

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Cookie.java [hunk 1]
```diff
@@ -42,12 +42,12 @@ public void deleteCookie(String name) {
...
 	public String getCookies() {
-		StringBuffer ret = new StringBuffer();
+		StringBuilder ret = new StringBuilder();
 
-		ArrayList<org.openqa.selenium.Cookie> cookies = new ArrayList<org.openqa.selenium.Cookie>(browserManagement
+		ArrayList<org.openqa.selenium.Cookie> cookies = new ArrayList<>(browserManagement
 				.getCurrentWebDriver().manage().getCookies());
 		for (int i = 0; i < cookies.size(); i++) {
-			ret.append(cookies.get(i).getName() + "=" + cookies.get(i).getValue());
+			ret.append(cookies.get(i).getName()).append("=").append(cookies.get(i).getValue());
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Cookie.java]: How about
```diff
-.append("=")
+.append('=')
```
? :)

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/JavaScript.java [hunk 2]
```diff
@@ -165,8 +165,7 @@ public Object executeAsyncJavascript(String... code) {
 	public String getAlertMessage() {
 		try {
 			Alert alert = browserManagement.getCurrentWebDriver().switchTo().alert();
-			String text = alert.getText().replace("\n", "");
-			return text;
+			return alert.getText().replace("\n", "");
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/JavaScript.java]: Can we use `.replace('\n', '')` instead?

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java [hunk 3]
```diff
@@ -222,7 +222,6 @@ public void selectFromList(String locator, String... items) {
 				} catch (NoSuchElementException e2) {
 					nonExistingItems.add(item);
 					lastItemFound = false;
-					continue;
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java]: I'd rather keep it. IMHO.

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java [hunk 4]
```diff
@@ -257,10 +256,7 @@ public void selectFromListByIndex(String locator, String... indexes) {
 			throw new SeleniumLibraryNonFatalException("No index given.");
 		}
 
-		List<String> tmp = new ArrayList<String>();
-		for (String index : indexes) {
-			tmp.add(index);
-		}
+		List<String> tmp = new ArrayList<>(Arrays.asList(indexes));
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java]: How about using `Collections.addAll()` instead?

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/JavaScript.java (hunk 2 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/JavaScript.java]: it is not working. only alert.getText().replace("\n", "");

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Cookie.java (hunk 1 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Cookie.java]: Done

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java [hunk 5]
```diff
@@ -52,14 +52,8 @@ public void waitForCondition(final String condition, String...params) {
...
-		waitUntil(timeout, message, new WaitUntilFunction() {
-
-			@Override
-			public boolean isFinished() {
-				return Boolean.TRUE.equals(((JavascriptExecutor) browserManagement.getCurrentWebDriver())
-						.executeScript(condition));
-			}
-		});
+		waitUntil(timeout, message, () -> Boolean.TRUE.equals(((JavascriptExecutor) browserManagement.getCurrentWebDriver())
+				.executeScript(condition)));
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java]: Could you, please, put `waitUntil()` argument to the separate lines to make code a little bit easier to read?

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java (hunk 4 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java]: Done

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java [hunk 6]
```diff
@@ -429,12 +343,12 @@ protected void waitUntil(String timestr, String message, WaitUntilFunction funct
...
 			try {
 				Thread.sleep(200);
-			} catch (InterruptedException e) {
+			} catch (InterruptedException ignored) {
 			}
 		}
 	}
 
-	protected static interface WaitUntilFunction {
+	protected interface WaitUntilFunction {
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java]: This method look like a candidate for having [`FunctionalInterface` annotation](https://docs.oracle.com/javase/8/docs/api/java/lang/FunctionalInterface.html).

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/ElementFinder.java [hunk 7]
```diff
@@ -301,33 +289,43 @@ protected static Strategy parseLocator(FindByCoordinates findByCoordinates, Stri
...
-			constraints.put("type", "checkbox");
-		} else if (tag.equals("text field")) {
-			tag = "input";
-			constraints.put("type", "text");
-		} else if (tag.equals("file upload")) {
-			tag = "input";
-			constraints.put("type", "file");
-		}
+		Map<String, String> constraints = new TreeMap<>();
+        switch (tag) {
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/ElementFinder.java]: There is something wrong with indentation.

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java [hunk 8]
```diff
@@ -67,23 +64,20 @@ public static WebElement findByRow(WebDriver webDriver, String tableLocator, int
...
 		}
 		return searchInLocators(webDriver, formattedLocators, content);
 	}
 
 	protected static void addLocatorSuffix(Map<String, List<String>> locatorSuffixesMap, String key, String... values) {
-		List<String> list = new ArrayList<String>();
-		for (String value : values) {
-			list.add(value);
-		}
+		List<String> list = new ArrayList<>(Arrays.asList(values));
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java]: Collections.addAll()?

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/WindowManager.java [hunk 9]
```diff
@@ -45,45 +45,24 @@ public void select(WebDriver webDriver, SelectCoordinates selectCoordinates) {
...
-
-					@Override
-					public boolean match(List<String> currentWindowInfo) {
-						return currentWindowInfo.get(WINDOW_INFO_INDEX_DOCUMENT_TITLE).trim().toLowerCase()
-								.equals(selectCoordinates.criteria.toLowerCase());
-					}
-
-				}, "Unable to locate window with title '" + selectCoordinates.criteria + "'");
+				selectMatching(webDriver, currentWindowInfo -> currentWindowInfo.get(WINDOW_INFO_INDEX_DOCUMENT_TITLE).trim().toLowerCase()
+						.equals(selectCoordinates.criteria.toLowerCase()), "Unable to locate window with title '" + selectCoordinates.criteria + "'");
```
Comment (php-coder) [inline src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/WindowManager.java]: Could you split the arguments and make it more readable, please?

Comment (php-coder) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/WindowManager.java]: And for the text 2 methods also...

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/JavaScript.java (hunk 2 above)
Comment (php-coder) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/JavaScript.java]: Yes, I see it now. Thanks for correcting me!

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java (hunk 3 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/SelectElement.java]: Roll back

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java (hunk 5 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java]: Done

Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java]: Done

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/ElementFinder.java (hunk 7 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/ElementFinder.java]: Done

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java (hunk 8 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java]: Done

Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java]: Could you explain to me why this is better?

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/WindowManager.java (hunk 9 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/WindowManager.java]: Done

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java (hunk 6 above)
Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/keywords/Waiting.java]: Yep) fixed

File: src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java (hunk 8 above)
Comment (php-coder) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java]: `Arrays.asList()` creates a new ArrayList under the hood. After that, we pass it to a ArrayList() constructor where we create an array (see `toArray()` there), and after that we copy this array to a destination array. So, here we've copied elements 2 times and produced unnecessary temporary objects (more work to GC). We can avoid this by using `Collections.addAll()`. That was my point.

Comment (YauheniPo) [reply src/main/java/com/github/markusbernhardt/seleniumlibrary/locators/TableElementFinder.java]: ok, thanks a lot!

**Comment by Hi-Fi** [comment]
Thank you, refactoring has been long time on the list (and I think there's still plenty of it left).

**Pull request merged**